File: backend-node/controllers/reminderController.js
import Reminder from '../models/Reminder.js';

export const getAllReminders = async (req, res) => {
  try {
    const reminders = await Reminder.find({ userId: req.query.userId });
    res.json(reminders);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const getReminderById = async (req, res) => {
  try {
    const reminder = await Reminder.findOne({ _id: req.params.id, userId: req.query.userId });
    if (!reminder) return res.status(404).json({ message: 'Reminder not found' });
    res.json(reminder);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const createReminder = async (req, res) => {
  try {
    const reminder = new Reminder(req.body);
    const savedReminder = await reminder.save();
    res.status(201).json(savedReminder);
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

export const updateReminder = async (req, res) => {
  try {
    const reminder = await Reminder.findOneAndUpdate(
      { _id: req.params.id, userId: req.query.userId },
      req.body,
      { new: true }
    );
    if (!reminder) return res.status(404).json({ message: 'Reminder not found' });
    res.json(reminder);
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

export const deleteReminder = async (req, res) => {
  try {
    const reminder = await Reminder.findOneAndDelete({ _id: req.params.id, userId: req.query.userId });
    if (!reminder) return res.status(404).json({ message: 'Reminder not found' });
    res.json({ message: 'Reminder deleted successfully' });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const toggleReminder = async (req, res) => {
  try {
    const reminder = await Reminder.findOne({ _id: req.params.id, userId: req.query.userId });
    if (!reminder) return res.status(404).json({ message: 'Reminder not found' });
    
    reminder.enabled = !reminder.enabled;
    await reminder.save();
    res.json(reminder);
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

export const markAsTaken = async (req, res) => {
  try {
    const reminder = await Reminder.findOneAndUpdate(
      { _id: req.params.id, userId: req.query.userId },
      { lastTaken: new Date() },
      { new: true }
    );
    if (!reminder) return res.status(404).json({ message: 'Reminder not found' });
    res.json(reminder);
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

export const undoTaken = async (req, res) => {
  try {
    const reminder = await Reminder.findOneAndUpdate(
      { _id: req.params.id, userId: req.query.userId },
      { lastTaken: null },
      { new: true }
    );
    if (!reminder) return res.status(404).json({ message: 'Reminder not found' });
    res.json(reminder);
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

File: backend/src/main/java/com/meditracker/backend/controller/ReminderController.java
package com.meditracker.backend.controller;

import com.meditracker.backend.model.Reminder;
import com.meditracker.backend.service.ReminderService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/reminders")
@CrossOrigin(origins = {"http://localhost:5173", "http://localhost:3000"}, allowCredentials = "true")
public class ReminderController {
    
    @Autowired
    private ReminderService reminderService;
    
    @GetMapping
    public ResponseEntity<List<Reminder>> getAllReminders(@RequestParam String userId) {
        List<Reminder> reminders = reminderService.getRemindersByUserId(userId);
        return ResponseEntity.ok(reminders);
    }
    
    @GetMapping("/{id}")
    public ResponseEntity<Reminder> getReminderById(@PathVariable String id, @RequestParam String userId) {
        Reminder reminder = reminderService.getReminderById(id, userId);
        return ResponseEntity.ok(reminder);
    }
    
    @PostMapping
    public ResponseEntity<Reminder> createReminder(@Valid @RequestBody Reminder reminder) {
        Reminder createdReminder = reminderService.createReminder(reminder);
        return ResponseEntity.status(HttpStatus.CREATED).body(createdReminder);
    }
    
    @PutMapping("/{id}")
    public ResponseEntity<Reminder> updateReminder(
            @PathVariable String id, 
            @Valid @RequestBody Reminder reminder,
            @RequestParam String userId) {
        Reminder updatedReminder = reminderService.updateReminder(id, reminder, userId);
        return ResponseEntity.ok(updatedReminder);
    }
    
    @DeleteMapping("/{id}")
    public ResponseEntity<Map<String, String>> deleteReminder(@PathVariable String id, @RequestParam String userId) {
        reminderService.deleteReminder(id, userId);
        return ResponseEntity.ok(Map.of("message", "Reminder deleted successfully"));
    }
    
    @PatchMapping("/{id}/toggle")
    public ResponseEntity<Reminder> toggleReminder(@PathVariable String id, @RequestParam String userId) {
        Reminder reminder = reminderService.toggleReminder(id, userId);
        return ResponseEntity.ok(reminder);
    }
    
    @PatchMapping("/{id}/taken")
    public ResponseEntity<Reminder> markAsTaken(@PathVariable String id, @RequestParam String userId) {
        Reminder reminder = reminderService.markAsTaken(id, userId);
        return ResponseEntity.ok(reminder);
    }
    
    @PatchMapping("/{id}/undo-taken")
    public ResponseEntity<Reminder> undoTaken(@PathVariable String id, @RequestParam String userId) {
        Reminder reminder = reminderService.undoTaken(id, userId);
        return ResponseEntity.ok(reminder);
    }
}